src/formats/apple-notes.ts

```diff
@@ -0,0 +1,303 @@
+		for (let f of noteFolders) await this.resolveFolder(f.Z_PK);
+		
+		const notes = await this.database.all`
+			SELECT
+				z_pk, zfolder, ztitle1 FROM ziccloudsyncingobject 
+			WHERE
+				z_ent = 11
+				AND ztitle1 IS NOT NULL
+				AND zfolder != ${this.trashedFolder}
+		`;
+		this.noteCount = notes.length;
+		
+		for (let n of notes) {
+			try { 
+				await this.resolveNote(n.Z_PK); 
+			}
+			catch (e) { 
+				this.ctx.reportFailed(n.ZTITLE1, e?.message); 
+			}
+		}
+	}
+	
+	async resolveAccount(id: number): Promise<void> {
+		if (!this.multiAccount && Object.keys(this.resolvedAccounts).length) {
+			this.multiAccount = true;
+		}
+		
+		const account = await this.database.get`
+			SELECT zname, zidentifier FROM ziccloudsyncingobject
+			WHERE z_ent = 13 AND z_pk = ${id}
+		`;
+			
+		this.resolvedAccounts[id] = {
+			name: account.ZNAME,
+			uuid: account.ZIDENTIFIER
+		};
+	}
+	
+	async resolveFolder(id: number): Promise<void> {
+		if (id in this.resolvedFiles) return;
+		
+		const folder = await this.database.get`
+			SELECT ztitle2, zparent, zidentifier, zfoldertype, zowner
+			FROM ziccloudsyncingobject
+			WHERE z_ent = 14 AND z_pk = ${id}
+		`;
+		let prefix;
+		
+		if (folder.ZFOLDERTYPE == ANFolderType.Smart) {
+			return;	
+		}
+		else if (!this.importTrashed && folder.ZFOLDERTYPE == ANFolderType.Trash) {
+			this.trashedFolder = id;
+			return;	
+		}
+		else if (folder.ZPARENT !== null) {
+			this.resolveFolder(folder.ZPARENT);
+			prefix = this.resolvedFiles[folder.ZPARENT].path + '/';
+		}
+		else if (this.multiAccount) {
+			const account = this.resolvedAccounts[folder.ZOWNER].name;
+			prefix = `${this.rootFolder.path}/${account}/`;
+		}
+		else {
+			prefix = `${this.rootFolder.path}/`;
+		}
+		
+		if (folder.ZIDENTIFIER !== 'DefaultFolder-CloudKit') {
+			//notes in the default "Notes" folder are placed in the main directory
+			prefix += folder.ZTITLE2;
+		}
+		
+		this.resolvedFiles[id] = await this.createFolders(prefix);
+	}
+	
+	async resolveNote(id: number): Promise<void> {
+		if (id in this.resolvedFiles) return;
+
+		const row = await this.database.get`
+			SELECT 
+				nd.z_pk, hex(nd.zdata) as zhexdata, zcso.ztitle1, zfolder, 
+				zcreationdate2, zcreationdate3, zmodificationdate1, zispasswordprotected
+			FROM
+				zicnotedata AS nd, 
+				(SELECT *, NULL AS zcreationdate3 FROM ziccloudsyncingobject) AS zcso 
+			WHERE
+				zcso.z_pk = nd.znote
+				AND zcso.z_pk = ${id}
+		`;
+		
+		if (row.ZISPASSWORDPROTECTED) {
+			this.ctx.reportSkipped(row.ZTITLE1, 'note is password protected');
+			return;
+		}
+		
+		const value = this.decodeData<ANDocument>(row.zhexdata).note;		
+		const folder = this.resolvedFiles[row.ZFOLDER] as TFolder;
+		const title = `${row.ZTITLE1}.md`;
+		
+		this.resolvedFiles[id] = await this.saveAsMarkdownFile(folder, title, ''); 
+		
+		//notes may reference other notes, so we want them in resolvedFiles before we parse to avoid cycles
```

Comment:
They are added to resolvedFiles before they are parsed (by creating a `NoteConverter`). What I mean here is that since note references are checked against resolvedFiles, a note that references itself (or a note that references another note, which references the first note) won't create a cycle